net: ping: Improve the output of the ping function

benpicco opened the pull request:
The output of the `net ping` function is a bit terse.

Let's make it more Linux-like by reporting TTL, measuring round-trip-time and also display RSSI when available.

**before:**
```
uart:~$ net ping fe80::7b7a:342f:9d73:b482
Sent a ping to fe80::7b7a:342f:9d73:b482
Received echo reply from fe80::7b7a:342f:9d73:b482 to fe80::d07:d5e9:7c9:572d
```

**after:**
```
uart:~$ net ping fe80::7b7a:342f:9d73:b482
PING fe80::7b7a:342f:9d73:b482
8 bytes from fe80::7b7a:342f:9d73:b482 to fe80::7841:4946:59b8:3d9c: icmp_seq=0 ttl=64 rssi=-53 dBm time=6.65ms
8 bytes from fe80::7b7a:342f:9d73:b482 to fe80::7841:4946:59b8:3d9c: icmp_seq=1 ttl=64 rssi=-57 dBm time=6.59ms
8 bytes from fe80::7b7a:342f:9d73:b482 to fe80::7841:4946:59b8:3d9c: icmp_seq=2 ttl=64 rssi=-57 dBm time=6.56ms
```

benpicco committed:
net/icmpv4: Include icmp_hdr in callback

Allow accessing already parsed information from the ICMP header
that callbacks might be interested in.

This makes the callback signature and behaviour match that of
the ICMPv6 implementation.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>

pfalcon commented on subsys/net/ip/net_shell.c:
```diff
@@ -2790,9 +2830,24 @@ static int cmd_net_ping(const struct shell *shell, size_t argc, char *argv[])
+	for (size_t i = 1; i < argc - 1; ++i) {
+		switch (*argv[i]) {
+		case 'c':
+			count = atoi(argv[i]+2);
```
Codestyle: space about operators.

pfalcon commented on subsys/net/ip/net_shell.c:
```diff
@@ -2790,9 +2830,24 @@ static int cmd_net_ping(const struct shell *shell, size_t argc, char *argv[])
+			count = atoi(argv[i]+2);
+			break;
+		case 'i':
+			interval = atoi(argv[i]+2);
```
Codestyle: space about operators.

pfalcon commented on subsys/net/ip/net_shell.c:
```diff
@@ -2700,16 +2714,23 @@ static int _ping_ipv6(const struct shell *shell, char *host)
+				sys_rand32_get(),
+				i);
+
+		if (ret)
```
Codestyle: braces are mandatory even around single statement.

pfalcon commented on subsys/net/ip/net_shell.c:
```diff
@@ -2790,9 +2830,24 @@ static int cmd_net_ping(const struct shell *shell, size_t argc, char *argv[])
+
+	for (size_t i = 1; i < argc - 1; ++i) {
+		switch (*argv[i]) {
+		case 'c':
```
I don't see commit message mentioning that new "ping" command options are added. Neither help text seems to be updated.

And as far as I can tell, the command would look like `ping c5 i1 host`. Why don't we make it truly similar to unix commands and be `ping -c5 -i1 host`.

jukkar replied:
I agree with @pfalcon here, use -c and -i for options, it might be also good idea to allow a space character between option and the value.

benpicco replied:
Ok, I initially wanted to avoid adding a dependency on libc or re-implementing `getopt()` poorly, but now have opted for the latter.

pfalcon reviewed (dismissed):
That looks very nice features, thanks!

But there're thing to elaborate in the patch still.

jukkar commented on subsys/net/ip/icmpv4.c:
```diff
@@ -208,6 +208,7 @@ int net_icmpv4_send_echo_request(struct net_if *iface,
 
```
In commit "net: ping: Improve the output of the ping function", I would instead name it "net: shell: Improve the output of the ping function" as this is more a shell improvement.

benpicco replied:
Ok, reworded the commit message.

jukkar reviewed (changes requested):
Thanks @benpicco, this looks really nice. Just few fixes and we are good to go.

tbursztyka commented on subsys/net/ip/icmpv4.c:
```diff
@@ -208,6 +208,7 @@ int net_icmpv4_send_echo_request(struct net_if *iface,
 
 	echo_req->identifier = htons(identifier);
 	echo_req->sequence   = htons(sequence);
+	echo_req->time_stamp = htonl(k_cycle_get_32());
```
so line 191 you would need to add sizeof(u32_t) in addition of the echo_req

then here, remove that line. And after line 213 below just: net_pkt_write_be32_new(k_cycle_get_32());

tbursztyka commented on subsys/net/ip/icmpv6.c:
```diff
@@ -332,6 +332,7 @@ int net_icmpv6_send_echo_request(struct net_if *iface,
 
 	echo_req->identifier = htons(identifier);
 	echo_req->sequence   = htons(sequence);
+	echo_req->time_stamp = htonl(k_cycle_get_32());
```
apply the same comment as in icmpv4.c

tbursztyka commented on subsys/net/ip/net_shell.c:
```diff
@@ -2659,24 +2659,38 @@ static enum net_verdict _handle_ipv6_echo_reply(struct net_pkt *pkt,
 						struct net_icmp_hdr *icmp_hdr)
 {
-	ARG_UNUSED(icmp_hdr);
+	struct net_icmpv6_echo_req *echo_repl = (struct net_icmpv6_echo_req *)(icmp_hdr + 1);
```
not good, see my comment in ipv4_echo_reply cb

benpicco replied:
Yea I didn't like that either.
So I understood you right, I'll change it so that the handler gets the `struct net_icmpv4_echo_req` instead of `struct net_icmp_hdr` .

benpicco replied:
nvm, I figured `NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE` / `net_pkt_get_data_new()` is the way to go.

tbursztyka commented on subsys/net/ip/net_shell.c:
```diff
@@ -2738,44 +2760,79 @@ static enum net_verdict _handle_ipv4_echo_reply(struct net_pkt *pkt,
+	PR("PING %s\n", host);
+
+	for (int i = 0; i < count; ++i) {
+
```
why the empty line?

benpicco replied:
I removed it.

tbursztyka commented on subsys/net/ip/net_shell.c:
```diff
@@ -2738,44 +2760,79 @@ static enum net_verdict _handle_ipv4_echo_reply(struct net_pkt *pkt,
+	for (int i = 0; i < count; ++i) {
+
+		ret = net_icmpv4_send_echo_request(
+				iface,
```
move iface right after the preceding '('

and indent properly the parameters below '(' + 1 space

benpicco replied:
moved it around

tbursztyka commented on subsys/net/ip/net_shell.c:
```diff
@@ -2738,44 +2760,79 @@ static enum net_verdict _handle_ipv4_echo_reply(struct net_pkt *pkt,
+				&ipv4_target,
+				sys_rand32_get(),
+				i);
+
```
no need of the extra empty line here as the if is going to check ret

benpicco replied:
I removed the empty line.

tbursztyka commented on subsys/net/ip/net_shell.c:
```diff
@@ -2738,44 +2760,79 @@ static enum net_verdict _handle_ipv4_echo_reply(struct net_pkt *pkt,
+				sys_rand32_get(),
+				i);
+
+		if (ret)
```
even one statement if needs the {} (see zephyr's code style)

tbursztyka reviewed (changes requested):
Split the 2nd commit in 3: 
- one adding timestamp as arbitrary data for icmpv4
- one adding timestamp as arbitrary data for icmpv6
- one improving shell 

and apply all the comments as well.

pfalcon commented:
@benpicco, Thanks for addressing the initial concerns raised.

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2659,24 +2659,38 @@ static enum net_verdict _handle_ipv6_echo_reply(struct net_pkt *pkt,
+		 ntohs(echo_repl->sequence),
+		 ip_hdr->hop_limit,
+#ifdef CONFIG_IEEE802154
+		 (s8_t) pkt->ieee802154_rssi,
```
The data coming from the [nrf52](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ieee802154/ieee802154_nrf5.h#L21) is a s8_t tough, and RSSI is usually given in negative dBm.

The only other use is currently in [openthread.c](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/l2/openthread/openthread.c#L226) where it is [implicitly cast to s8_t again](https://github.com/openthread/openthread/blob/master/include/openthread/platform/radio.h#L183).

I think this is a bug in Zephyr's ieee802154 API.

tbursztyka replied:
zephyr's 15.4 stack implements rssi as specified in 15.4 specs (8bits value from 0 to 255).
Ref: spec 2012 chapter 18.2.3.3

Though dbm would be indeed better (as it is the scale used also in BT and WiFi). I don't know why 15.4 came with its own definition.

benpicco replied:
Hm, with two's complement it doesn't make a difference either way, higher values are better no matter if they are interpreted as signed or not.

If the standard says so I'll remove the cast.

benpicco commented on subsys/net/ip/icmpv4.h:
```diff
@@ -30,6 +30,7 @@
 struct net_icmpv4_echo_req {
 	u16_t identifier;
 	u16_t sequence;
+	u32_t time_stamp;
```
Yea you are right. I've changed it to a `u8_t data` and let the application decide what arbitrary payload to put in there (if any).

benpicco committed:
net/icmpv4: Allow for arbitrary payload data in ICMP echo

Allow for including arbitrary data in net_icmpv4_send_echo_request()
that will be echoed verbatim by the receiver.

This allows to use ICMP echo for diagnostic use cases, e.g. by testing
packet framentation (large payload) or measuring round-trip-time.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
net/icmpv6: Allow for arbitrary payload data in ICMP echo

Allow for including arbitrary data in net_icmpv6_send_echo_request()
that will be echoed verbatim by the receiver.

This allows to use ICMP echo for diagnostic use cases, e.g. by testing
packet framentation (large payload) or measuring round-trip-time.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco committed:
net: shell: Improve the output of the ping function

Report rtt, ttl and rssi if available.

Signed-off-by: Benjamin Valentin <benjamin.valentin@ml-pa.com>

benpicco commented on doc/reference/networking/net_shell.rst:
```diff
@@ -34,7 +34,7 @@ The following net-shell commands are implemented:
    "net nbr", "Print neighbor information. Only available if
    :option:`CONFIG_NET_IPV6` is set."
-   "net ping", "Ping a network host."
+   "net ping [-c count] [-i interval]", "Ping a network host."
```
fixed

dbkinder reviewed (approved):
+1 for doc change

tbursztyka commented on subsys/net/ip/icmpv4.h:
```diff
@@ -30,6 +30,7 @@
 struct net_icmpv4_echo_req {
 	u16_t identifier;
 	u16_t sequence;
+	u8_t  data[];
```
So that won't be necessary then

benpicco replied:
removed

tbursztyka commented on subsys/net/ip/icmpv4.c:
```diff
@@ -208,6 +210,7 @@ int net_icmpv4_send_echo_request(struct net_if *iface,
 
 	echo_req->identifier = htons(identifier);
 	echo_req->sequence   = htons(sequence);
+	memcpy(echo_req->data, data, data_size);
```
Even though, at this stage of this function implementation, this logic is not wrong, it won't fly with time if one day the icmpv4_access is not a contiguous one anymore.

And in any case, never memcpy (or memmove  etc..) by yourself. Alway use the net_pkt r/w api, because that API can tell you if it could r/w or not (it verifies that buffer does have the necessary space)

So after net_pkt_set_data() below, just do a net_pkt_write(pkt, data, data_size)

benpicco replied:
Ok, changed it to net_pkt_write_new().

tbursztyka commented on subsys/net/ip/icmpv6.c:
```diff
@@ -332,6 +334,7 @@ int net_icmpv6_send_echo_request(struct net_if *iface,
 
 	echo_req->identifier = htons(identifier);
 	echo_req->sequence   = htons(sequence);
+	memcpy(echo_req->data, data, data_size);
```
apply the same as for icmpv4.c

benpicco replied:
ok

tbursztyka commented on subsys/net/ip/icmpv4.c:
```diff
@@ -188,7 +190,7 @@ int net_icmpv4_send_echo_request(struct net_if *iface,
 
 	pkt = net_pkt_alloc_with_buffer(iface,
-					sizeof(struct net_icmpv4_echo_req),
+					sizeof(struct net_icmpv4_echo_req) + data_size,
```
80 chars limit

benpicco replied:
moved it to the next line

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2659,24 +2659,44 @@ static enum net_verdict _handle_ipv6_echo_reply(struct net_pkt *pkt,
+	struct net_icmpv6_echo_req *icmp_echo;
+	icmp_echo = (struct net_icmpv6_echo_req *)net_pkt_get_data_new(pkt,
+								       &icmp_access);
+	u32_t cycles;
```
Ok.

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2659,24 +2659,44 @@ static enum net_verdict _handle_ipv6_echo_reply(struct net_pkt *pkt,
+	NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE(icmp_access,
+					      struct net_icmpv6_echo_req);
+	struct net_icmpv6_echo_req *icmp_echo;
+	icmp_echo = (struct net_icmpv6_echo_req *)net_pkt_get_data_new(pkt,
```
Ok.

benpicco commented on subsys/net/ip/icmpv6.c:
```diff
@@ -312,7 +314,7 @@ int net_icmpv6_send_echo_request(struct net_if *iface,
 
 	pkt = net_pkt_alloc_with_buffer(iface,
-					sizeof(struct net_icmpv6_echo_req),
+					sizeof(struct net_icmpv6_echo_req) + data_size,
```
moved it to the next line

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2740,46 +2767,86 @@ static enum net_verdict _handle_ipv4_echo_reply(struct net_pkt *pkt,
+	struct net_icmpv6_echo_req *icmp_echo;
+	icmp_echo = (struct net_icmpv4_echo_req *)net_pkt_get_data_new(pkt,
+								       &icmp_access);
+	u32_t cycles;
```
ok

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2659,24 +2659,44 @@ static enum net_verdict _handle_ipv6_echo_reply(struct net_pkt *pkt,
+	icmp_echo = (struct net_icmpv6_echo_req *)net_pkt_get_data_new(pkt,
+								       &icmp_access);
+	u32_t cycles;
+	memcpy(&cycles, icmp_echo->data, sizeof(cycles));
```
Ok, changed it to use net_pkt_read_be32_new() instead.

benpicco commented:
I know, just thought it makes reviewing easier this way.
-> squashed.

jukkar reviewed (approved):
LGTM

benpicco commented:
@tbursztyka does this still need any changes?

tbursztyka reviewed (approved):
Seems fine to me

bublover commented:
Could you please add size parameter? 
ping host -s 1470

benpicco commented:
@bublover Ok, added it.

I'm not so happy about the VLA use, but `k_malloc` is only available when `CONFIG_HEAP_MEM_POOL_SIZE` is set - and that defaults to 0.

I've only tested the IPv6 version, but IPv4 should work the same.

bublover commented on subsys/net/ip/net_shell.c:
```diff
@@ -2698,12 +2718,14 @@ static enum net_verdict handle_ipv6_echo_reply(struct net_pkt *pkt,
 	struct in6_addr ipv6_target;
 	struct net_if *iface = net_if_get_default();
 	struct net_nbr *nbr;
+	uint8_t payload[size += sizeof(u32_t)];
```
Hi @benpicco ,
About this payload, is it allocated from heap or thread stack or something else?
If from thread stack, is it possible to be overflow with big size?

jukkar replied:
Using VLAs is no good, please do it differently. One option could be to allocate enough net_buf's for a given size, and then pass that chain of net_buf's directly to `net_icmpv6_send_echo_request()` which can then send that data directly.

benpicco replied:
Yes it's allocated on the stack and yes that's bad.

I'm not sure how to do it with net_bufs though. Is there a way where I don't have to reserve dedicated memory for it using `NET_BUF_POOL_DEFINE` but instead can make use of some existing pool?

Another way would be to only enable the -s option if `CONFIG_HEAP_MEM_POOL_SIZE` is not 0.

jukkar replied:
Hmm, we seem to have lost proper API in net_pkt.h for allocating net_buf's.

Another option could be to use `net_pkt_skip()` and optionally with `net_pkt_memset()`. That way we could fill enough net_buf's with given byte value. This would be needed to be done inside `net_icmpv6_send_echo_request()` as the net_pkt is not exposed outside of that function.

So for example we could have something like `net_icmpv6_send_echo_request(iface, dst, id, seq, len, value)`, and inside that func just do `net_pkt_skip(pkt, len)` followed by cursor setting and then `net_pkt_memset(pkt, value)`. The cursor setting is a bit vague here as I am just brainstorming atm.

jukkar replied:
After looking how net_pkt_skip() and net_pkt_memset() are implemented, it is probably enough just to call net_pkt_memset() which should allocate enough net_buf's to fill the packet.

jukkar commented:
@benpicco could you rebase against latest master and re-submit without the size support, we can add that one later in separate PR.

benpicco commented:
@jukkar alright!

easonxiang commented on subsys/net/ip/net_shell.c:
```diff
@@ -2734,46 +2773,98 @@ static inline void remove_ipv4_ping_handler(void)
+
+	cycles = k_cycle_get_32() - cycles;
+
+	PR_SHELL(shell_for_ping, "%d bytes from %s to %s: icmp_seq=%d ttl=%d time=%.2f ms\n",
```
Print float data depends on some additional configs, otherwise, the time can not output correctly.

28 bytes from 192.168.1.1 to 192.168.1.100: icmp_seq=1 ttl=64 time=%.2f ms

benpicco replied:
I suppose it's `CONFIG_NEWLIB_LIBC_FLOAT_PRINTF` and it only affects the formatting to two significant digits as `%f` is used elsewhere in that file.

Does it work for you now?

easonxiang replied:
It also doesn't work with %f, can you please use %d instead as this is a universal network tool.

28 bytes from 192.168.1.1 to 192.168.1.100: icmp_seq=0 ttl=64 time=%f ms

benpicco replied:
Ok.

benpicco replied:
What target / config are you using?
When I build for nrf52840_pca10056, `CONFIG_NEWLIB_LIBC_FLOAT_PRINTF` is not set, yet `%.2f` is available.

easonxiang replied:
@benpicco I'm using 96boards IVY5661, the board support patch(https://github.com/zephyrproject-rtos/zephyr/pull/10626) is not merged yet.

I tried to store time in a variable before output, It's very strange that printf works, but PR_SHELL not. It seems there is some issue with PR_SHELL.
```
    float time;
    time =  SYS_CLOCK_HW_CYCLES_TO_NS(cycles) / 1000000.f;
    printf("time: %f.\n", time);

    PR_SHELL(shell_for_ping, "%d bytes from %s to %s: icmp_seq=%d ttl=%d time=%f ms\n",
         ntohs(ip_hdr->len) - net_pkt_ipv6_ext_len(pkt) - NET_ICMPH_LEN,
         net_sprint_ipv4_addr(&ip_hdr->src),
         net_sprint_ipv4_addr(&ip_hdr->dst),
         ntohs(icmp_echo->sequence),
         ip_hdr->ttl,
         time);
```
OUTPUT:
time: 4.418798.
28 bytes from 192.168.1.1 to 192.168.1.101: icmp_seq=0 ttl=64 time=%f ms

easonxiang commented on subsys/net/ip/net_shell.c:
```diff
@@ -2659,24 +2659,52 @@ static enum net_verdict handle_ipv6_echo_reply(struct net_pkt *pkt,
+#endif
+		 SYS_CLOCK_HW_CYCLES_TO_NS(cycles) / 1000000.f);
 	k_sem_give(&ping_timeout);
-	remove_ipv6_ping_handler();
```
remove_ipv4_ping_handler should be removed also.

benpicco replied:
good catch!

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2740,46 +2776,95 @@ static enum net_verdict handle_ipv4_echo_reply(struct net_pkt *pkt,
+	const char *val = argv[*i] + 2;
+
+	if (*val) {
+		return atoi(val);
```
Ok, converted it to use `strtol()`

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2791,12 +2876,44 @@ static int cmd_net_ping(const struct shell *shell, size_t argc, char *argv[])
+			break;
+		case 'i':
+			interval = _parse_arg(&i, argc, argv);
+
```
removed

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2791,12 +2876,44 @@ static int cmd_net_ping(const struct shell *shell, size_t argc, char *argv[])
+		switch (argv[i][1]) {
+		case 'c':
+			count = _parse_arg(&i, argc, argv);
+
```
removed

benpicco commented on doc/reference/networking/net_shell.rst:
```diff
@@ -31,7 +31,7 @@ The following net-shell commands are implemented:
    "net nbr", "Print neighbor information. Only available if
    :option:`CONFIG_NET_IPV6` is set."
-   "net ping", "Ping a network host."
+   "net ping [-c count] [-i interval]", "Ping a network host."
```
Ok.

jukkar commented:
We have now another problem. When compiling `samples/net/sockets/echo_server` in qemu_x86, I get following fault:
```
uart:~$ net ping 2001:db8::2
PING 2001:db8::2
***** Floating point unit not enabled

Current thread ID = 0x004016f8
eax: 0x00403566, ebx: 0x00406558, ecx: 0x00000000, edx: 0x00000000
esi: 0x0040354e, edi: 0x0040357a, ebp: 0x00406a88, esp: 0x00406a50
eflags: 0x00000206 cs: 0x0008
call trace:
eip: 0x00015911
     0x000123ac (0x123ac)
     0x00012abc (0x406558)
     0x0000de38 (0x406558)
     0x0000de97 (0x1)
     0x0000479c (0x40655c)
     0x0000475a (0x4016e0)
Fatal fault in thread 0x004016f8! Aborting.
```

After enabling `CONFIG_FLOAT`, the output print looks like this
```
8 bytes from 2001:db8::2 to 2001:db8::1: icmp_seq=0 ttl=64 time=26113401 ms
8 bytes from 2001:db8::2 to 2001:db8::1: icmp_seq=1 ttl=64 time=-1608035756 ms
8 bytes from 2001:db8::2 to 2001:db8::1: icmp_seq=2 ttl=64 time=669327703 ms
```

You need to add () when dividing like this
```
diff --git a/subsys/net/ip/net_shell.c b/subsys/net/ip/net_shell.c
index bdb3e6ea4f..7a1836aef0 100644
--- a/subsys/net/ip/net_shell.c
+++ b/subsys/net/ip/net_shell.c
@@ -2696,7 +2696,7 @@ static enum net_verdict handle_ipv6_echo_reply(struct net_pkt *pkt,
 #ifndef CONFIG_NEWLIB_LIBC_FLOAT_PRINTF
                 (int)
 #endif
-                SYS_CLOCK_HW_CYCLES_TO_NS(cycles) / 1000000.f);
+                (SYS_CLOCK_HW_CYCLES_TO_NS(cycles) / 1000000.f));
        k_sem_give(&ping_timeout);
 
        net_pkt_unref(pkt);
```

For the float issue, add this change
```
index 7798d5a2f2..ca04d3e3c0 100644
--- a/subsys/net/ip/Kconfig
+++ b/subsys/net/ip/Kconfig
@@ -56,6 +56,7 @@ source "subsys/net/ip/Kconfig.ipv4"
 config NET_SHELL
        bool "Enable network shell utilities"
        select SHELL
+       select FLOAT
        help
          Activate shell module that provides network commands like
          ping to the console.
```

pfalcon commented:
> We have now another problem.
> ***** Floating point unit not enabled

So, I'd say we should keep avoiding floating point like we did before, and use fixed points calculations.

benpicco commented:
I've changed it so that it only uses float when `CONFIG_FLOAT` is set, otherwise it uses int and discards the sub-ms times.

@easonxiang does this also remedy the problem on uwp5661?

easonxiang commented:
> I've changed it so that it only uses float when `CONFIG_FLOAT` is set, otherwise it uses int and discards the sub-ms precision.
> 
> @easonxiang does this also remedy the problem on uwp5661?

@benpicco Yes, the int type works for me. I can not enable CONFIG_FLOAT as it depends on CPU_HAS_FPU, which is not available by uwp5661. 
Thanks for your quickly response.

jukkar reviewed (changes requested):
We are almost there, just one tweak needed.

benpicco commented on subsys/net/ip/net_shell.c:
```diff
@@ -2740,46 +2784,111 @@ static enum net_verdict handle_ipv4_echo_reply(struct net_pkt *pkt,
+
+	cycles = k_cycle_get_32() - cycles;
+
+	PR_SHELL(shell_for_ping, "%d bytes from %s to %s: icmp_seq=%d ttl=%d"
```
Good catch! I don't have a setup yet for testing legacy IP :wink:
